webauthn-authenticator-rs/src/MacAuthn/Sources/MacAuthn/MacAuthn.swift

```diff
@@ -0,0 +1,194 @@
+        let transports = credential.transports?.map {
+            ASAuthorizationSecurityKeyPublicKeyCredentialDescriptor.Transport.init(rawValue: $0)
+        } ?? ASAuthorizationSecurityKeyPublicKeyCredentialDescriptor.Transport.allSupported
+        let credential = ASAuthorizationSecurityKeyPublicKeyCredentialDescriptor(credentialID: id, transports: transports)
+        securityKeyRequest.excludedCredentials.append(credential)
+    }
+    
+    securityKeyRequest.attestationPreference = ASAuthorizationPublicKeyCredentialAttestationKind(rawValue: options.attestation ?? "none")
+    securityKeyRequest.userVerificationPreference = ASAuthorizationPublicKeyCredentialUserVerificationPreference.init(rawValue: options.authenticatorSelection.userVerification ?? "preferred")
+    
+    if options.authenticatorSelection.requireResidentKey == true {
+        securityKeyRequest.residentKeyPreference = .required
+    } else {
+        securityKeyRequest.residentKeyPreference = .preferred
+    }
+    
+    let authController = ASAuthorizationController(authorizationRequests: [platformKeyRequest, securityKeyRequest])
+    
+    return SRString(run(authController: authController))
+}
+
+@_cdecl("perform_auth")
+public func performAuth(options: SRString) -> SRString {
+    let options = try! JSONDecoder().decode(PublicKeyCredentialRequestOptions.self, from: Data(options.toArray()))
+    
+    let securityKeyProvider = ASAuthorizationSecurityKeyPublicKeyCredentialProvider(relyingPartyIdentifier: options.rpId)
+    let securityKeyRequest = securityKeyProvider.createCredentialAssertionRequest(challenge: options.challenge.decodeBase64Url()!)
+    
+    let platformProvider = ASAuthorizationPlatformPublicKeyCredentialProvider(relyingPartyIdentifier: options.rpId)
+    let platformKeyRequest = platformProvider.createCredentialAssertionRequest(challenge: options.challenge.decodeBase64Url()!)
+    
+    securityKeyRequest.userVerificationPreference = ASAuthorizationPublicKeyCredentialUserVerificationPreference.init(rawValue: options.userVerification ?? "preferred")
+    
+    securityKeyRequest.allowedCredentials = []
+    for credential in (options.allowCredentials ?? []) {
+        let id = credential.id.decodeBase64Url()!
+        let transports = credential.transports?.map {
+            ASAuthorizationSecurityKeyPublicKeyCredentialDescriptor.Transport.init(rawValue: $0)
+        } ?? ASAuthorizationSecurityKeyPublicKeyCredentialDescriptor.Transport.allSupported
+        let descriptor = ASAuthorizationSecurityKeyPublicKeyCredentialDescriptor(credentialID: id, transports: transports)
+        securityKeyRequest.allowedCredentials.append(descriptor)
+    }
+    // Setting allowedCredentials can hang for some reason: https://developer.apple.com/forums/thread/727267
+    securityKeyRequest.allowedCredentials = []
+    
+    let authController = ASAuthorizationController(authorizationRequests: [platformKeyRequest, securityKeyRequest])
```

Comment:
Given that `ASAuthorizationSecurityKeyPublicKeyCredentialProvider` is extremely limited (compared to Windows), I feel like this should be optional (ie: default to both, but use `authenticator_attachment` to select which backends it tries to use).